backends/dpdk/dpdkArch.cpp

```diff
@@ -2877,4 +2879,174 @@ const IR::Node* ElimHeaderCopy::postorder(IR::Member* m) {
     return m;
 }
 
+const IR::Node* DpdkAddPseudoHeaderDecl::preorder(IR::P4Program* program) {
+    if (is_all_args_header) return program;
+    auto* annotations = new IR::Annotations({new IR::Annotation(IR::ID("__pseudo_header__"), {})});
+    const IR::Type_Header* pseudo_hdr = new IR::Type_Header(headerTypeName, annotations);
+    allTypeDecls.push_back(pseudo_hdr);
+    allTypeDecls.append(program->objects);
+    program->objects = allTypeDecls;
+    return program;
+}
+
+// add pseudo header field to the headers struct for it be instantiated
+const IR::Node* DpdkAddPseudoHeaderDecl::preorder(IR::Type_Struct* st) {
+    if (is_all_args_header) return st;
+    bool header_found = isHeadersStruct(st);
+    if (header_found) {
+        IR::IndexedVector<IR::StructField> fields;
+        auto* annotations =
+            new IR::Annotations({new IR::Annotation(IR::ID("__pseudo_header__"), {})});
+        auto* type = new IR::Type_Name(new IR::Path(headerTypeName));
+        const IR::StructField* new_field1 =
+            new IR::StructField(headerInstanceName, annotations, type);
+        fields = st->fields;
+        fields.push_back(new_field1);
+        auto* st1 =
+            new IR::Type_Struct(st->srcInfo, st->name, st->annotations, st->typeParameters, fields);
+        return st1;
+    }
+    return st;
+}
+
+bool DpdkAddPseudoHeaderDecl::isHeadersStruct(const IR::Type_Struct* st) {
+    if (!st) return false;
+    auto annon = st->getAnnotation("__packet_data__");
+    if (annon == nullptr) {
+        if (structure->header_type == st->name.name)
+            return true;
+        else {
```

Comment:
The annotations are added early on to mark the main header struct and should be enough to detect this structure.

Reply:
done